iis-wf/iis-wf-export-actionmanager/src/main/java/eu/dnetlib/iis/wf/export/actionmanager/module/CitationsActionBuilderModuleUtils.java

```diff
@@ -1,56 +1,132 @@
-            if (result.getIdentifiers() == null) {
-                result.setIdentifiers(new ArrayList<TypedId>());
+    }
+
+    /**
+     * Build identifiers value of blob citation entry from citation entry.
+     */
+    public static class IdentifiersBuilder {
+        public static List<TypedId> build(CitationEntry citationEntry) {
+            return build(citationEntry,
+                    TypedIdFromDestinationDocumentIdBuilder::isValid,
+                    TypedIdFromDestinationDocumentIdBuilder::build,
+                    TypedIdsFromExternalDestinationDocumentIdsBuilder::isValid,
+                    TypedIdsFromExternalDestinationDocumentIdsBuilder::build);
+        }
+
+        public static List<TypedId> build(CitationEntry citationEntry,
+                                          Function<CitationEntry, Boolean> destinationDocumentIdValidator,
+                                          Function<CitationEntry, TypedId> typedIdFromDestinationDocumentIdFn,
+                                          Function<CitationEntry, Boolean> externalDestinationDocumentIdsValidator,
+                                          Function<CitationEntry, List<TypedId>> typedIdsFromExternalDestinationDocumentIdsFn) {
+            List<TypedId> identifiers = new ArrayList<>();
+            if (destinationDocumentIdValidator.apply(citationEntry)) {
+                identifiers.add(typedIdFromDestinationDocumentIdFn.apply(citationEntry));
+            }
+            if (externalDestinationDocumentIdsValidator.apply(citationEntry)) {
+                identifiers.addAll(typedIdsFromExternalDestinationDocumentIdsFn.apply(citationEntry));
+            }
+            return identifiers.isEmpty() ? null : identifiers;
+        }
+
+        public static class TypedIdFromDestinationDocumentIdBuilder {
+            public static Boolean isValid(CitationEntry citationEntry) {
+                return Objects.nonNull(citationEntry.getDestinationDocumentId());
             }
-            for (Entry<CharSequence, CharSequence> extId : entry.getExternalDestinationDocumentIds().entrySet()) {
-                result.getIdentifiers().add(new TypedId(extId.getValue().toString(), extId.getKey().toString(),
-                        1f * InfoSpaceConstants.CONFIDENCE_TO_TRUST_LEVEL_FACTOR));
+
+            public static TypedId build(CitationEntry citationEntry) {
+                return build(citationEntry, ConfidenceLevelBuilder::build);
+            }
+
+            public static TypedId build(CitationEntry citationEntry, Function<CitationEntry, Float> confidenceLevelFn) {
+                return new TypedId(citationEntry.getDestinationDocumentId().toString(),
+                        ExtraInfoConstants.CITATION_TYPE_OPENAIRE,
+                        confidenceLevelFn.apply(citationEntry));
+            }
+
+            public static class ConfidenceLevelBuilder {
```

Comment:
I guess we should rename it to `TrustLevelBuilder` because it builds trust level in the end...

Reply:
This was my initial name also, but this class sets `confidenceLevel` field in `TypedId` class so to avoid any misunderstandings I took the name from the field.

Reply:
OK, this is another advantage of doing code reviews: finding the bugs in existing code. We should either rename it within the XML blob (within TypedId) or skip applying the multiplier to reduce it to trust level.

I will create a dedicated ticket to handle this, no need to take care of it here.